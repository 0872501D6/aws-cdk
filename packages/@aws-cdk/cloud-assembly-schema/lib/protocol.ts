import * as fs from 'fs';
import { Schema, Validator } from 'jsonschema';
import { ArtifactMetadataEntryType, ArtifactType, AssemblyManifest } from './assembly-manifest';

/**
 * Protocol utility class.
 */
export class Manifest {

    /**
     * Save manifest to file.
     *
     * @param manifest - manifest.
     */
    public static save(manifest: AssemblyManifest, filePath: string) {
        fs.writeFileSync(filePath, JSON.stringify(manifest, undefined, 2));
    }

    /**
     * Load manifest from file.
     */
    public static load(filePath: string): AssemblyManifest {
        const raw: AssemblyManifest = JSON.parse(fs.readFileSync(filePath, 'UTF-8'));
        const manifest: AssemblyManifest = Manifest.validate(raw);
        return manifest;
    }

    /**
     * Fetch the current schema version number.
     */
    public static version(): string {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        return require('../schema/cloud-assembly.version.json').version;
    }

    // eslint-disable-next-line @typescript-eslint/no-require-imports
    private static schema: Schema = require('../schema/cloud-assembly.schema.json');

    private static validate(manifest: any): AssemblyManifest {
        const validator = new Validator();
        const result = validator.validate(manifest, Manifest.schema, {

            // does exist but is not in the TypeScript definitions
            nestedErrors: true,

            allowUnknownAttributes: false

        } as any);
        if (result.valid) { return manifest; }
        throw new Error(`Invalid assembly manifest:\n${result}`);
    }

    /**
     * This requires some explaining...
     *
     * We previously used `{ Key, Value }` for the object that represents a stack tag. (Notice the casing)
     * @link https://github.com/aws/aws-cdk/blob/v1.27.0/packages/aws-cdk/lib/api/cxapp/stacks.ts#L427.
     *
     * When that object moved to this package, it had to be JSII compliant, which meant the property
     * names must be `camelCased`, and not `PascalCased`. This meant it no longer matches the structure in the `manifest.json` file.
     * In order to support current manifest files, we have to translate the `PascalCased` representation to the new `camelCased` one.
     *
     * Note that the serialization itself still writes `PascalCased` because it relates to how CloudFormation expects it.
     *
     * Ideally, we would start writing the `camelCased` and translate to how CloudFormation expects it when needed. But this requires nasty
     * backwards-compatibility code and it just doesn't seem to be worth the effort.
     */
    private static patchStackTags(manifest: AssemblyManifest) {
        for (const artifact of Object.values(manifest.artifacts || [])) {
            if (artifact.type === ArtifactType.AWS_CLOUDFORMATION_STACK) {
                for (const metadataEntries of Object.values(artifact.metadata || [])) {
                    for (const metadataEntry of metadataEntries) {
                        if (metadataEntry.type === ArtifactMetadataEntryType.STACK_TAGS && metadataEntry.data) {

                            const metadataAny = metadataEntry as any;

                            metadataAny.data = metadataAny.data.map((t: any) => {
                                return { key: t.Key, value: t.Value };
                            });
                        }
                    }
                }
            }
        }
    }

    private constructor() {}

}
