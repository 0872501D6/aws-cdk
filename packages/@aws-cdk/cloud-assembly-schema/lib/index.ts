export * from './assembly-manifest';
export * from './protocol';
