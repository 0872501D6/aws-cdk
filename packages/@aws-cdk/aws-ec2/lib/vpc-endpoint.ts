import iam = require('@aws-cdk/aws-iam');
import cdk = require('@aws-cdk/cdk');
import { Connections, IConnectable } from './connections';
import { CfnVPCEndpoint } from './ec2.generated';
import { SecurityGroup } from './security-group';
import { VpcSubnet } from './vpc';
import { IVpcNetwork, IVpcSubnet, SubnetSelection } from './vpc-ref';

/**
 * A VPC endpoint.
 */
export interface IVpcEndpoint extends cdk.IConstruct {
  /**
   * The VPC endpoint identifier.
   */
  readonly vpcEndpointId: string;
}

/**
 * A VPC gateway endpoint.
 */
export interface IVpcGatewayEndpoint extends IVpcEndpoint {
  /**
   * Exports this VPC endpoint from the stack.
   */
  export(): VpcGatewayEndpointImportProps;
}

/**
 * The type of VPC endpoint.
 */
export enum VpcEndpointType {
  /**
   * Interface
   */
  Interface = 'Interface',

  /**
   * Gateway
   */
  Gateway = 'Gateway'
}

/**
 * A VPC endpoint service.
 */
export class VpcEndpointService {
  constructor(public readonly name: string, public readonly type: VpcEndpointType) {}
}

/**
 * A VPC endpoint AWS service.
 */
export class VpcEndpointAwsService extends VpcEndpointService {
  public static readonly SageMakeNotebook = new VpcEndpointAwsService('sagemaker', VpcEndpointType.Interface, 'aws.sagemaker');
  public static readonly CloudFormation = new VpcEndpointAwsService('cloudformation');
  public static readonly CloudTrail = new VpcEndpointAwsService('cloudtrail');
  public static readonly CodeBuild = new VpcEndpointAwsService('codebuild');
  public static readonly CodeBuildFips = new VpcEndpointAwsService('codebuil-fips');
  public static readonly CodeCommit = new VpcEndpointAwsService('codecommit');
  public static readonly CodeCommitFips = new VpcEndpointAwsService('codecommit-fips');
  public static readonly CodePipeline = new VpcEndpointAwsService('codepipeline');
  public static readonly Config = new VpcEndpointAwsService('config');
  public static readonly DynamoDb = new VpcEndpointAwsService('dynamodb', VpcEndpointType.Gateway);
  public static readonly Ec2 = new VpcEndpointAwsService('ec2');
  public static readonly Ec2Messages = new VpcEndpointAwsService('ec2messages');
  public static readonly Ecr = new VpcEndpointAwsService('ecr.api');
  public static readonly EcrDocker = new VpcEndpointAwsService('ecr.dkr');
  public static readonly Ecs = new VpcEndpointAwsService('ecs');
  public static readonly EcsAgent = new VpcEndpointAwsService('ecs-agent');
  public static readonly EcsTelemetry = new VpcEndpointAwsService('ecs-telemetry');
  public static readonly ElasticInferenceRuntime = new VpcEndpointAwsService('elastic-inference.runtime');
  public static readonly ElasticLoadBalancing = new VpcEndpointAwsService('elasticloadbalancing');
  public static readonly CloudWatchEvents = new VpcEndpointAwsService('events');
  public static readonly ApiGateway = new VpcEndpointAwsService('execute-api');
  public static readonly CodeCommitGit = new VpcEndpointAwsService('git-codecommit');
  public static readonly CodeCommitGitFips = new VpcEndpointAwsService('git-codecommit-fips');
  public static readonly KinesisStreams = new VpcEndpointAwsService('kinesis-streams');
  public static readonly Kms = new VpcEndpointAwsService('kms');
  public static readonly CloudWatchLogs = new VpcEndpointAwsService('logs');
  public static readonly CloudWatch = new VpcEndpointAwsService('monitoring');
  public static readonly S3 = new VpcEndpointAwsService('s3', VpcEndpointType.Gateway);
  public static readonly SageMakerApi = new VpcEndpointAwsService('sagemaker.api');
  public static readonly SageMakerRuntime = new VpcEndpointAwsService('sagemaker.runtime');
  public static readonly SageMakerRuntimeFips = new VpcEndpointAwsService('sagemaker.runtime-fips');
  public static readonly SecretsManager = new VpcEndpointAwsService('secretsmanager');
  public static readonly ServiceCatalog = new VpcEndpointAwsService('servicecatalog');
  public static readonly Sns = new VpcEndpointAwsService('sns');
  public static readonly Sqs = new VpcEndpointAwsService('sqs');
  public static readonly Ssm = new VpcEndpointAwsService('ssm');
  public static readonly SsmMessages = new VpcEndpointAwsService('ssmmessages');
  public static readonly Sts = new VpcEndpointAwsService('sts');
  public static readonly Transfer = new VpcEndpointAwsService('transfer.server');

  constructor(name: string, type?: VpcEndpointType, prefix?: string) {
    super(`${prefix || 'com.amazonaws'}.${cdk.Aws.region}.${name}`, type || VpcEndpointType.Interface);
  }
}

/**
 * Options to add an endpoint to a VPC.
 */
export interface VpcEndpointOptions {
  /**
   * The name of the service.
   */
  service: VpcEndpointService;
}

/**
 * Options to add a gateway endpoint to a VPC.
 */
export interface VpcGatewayEndpointOptions extends VpcEndpointOptions {
  /**
   * Where to add endpoint routing.
   *
   * @default private subnets
   */
  subnets?: SubnetSelection[]
}

/**
 * Construction properties for a GatewayEndpoint.
 */
export interface VpcGatewayEndpointProps extends VpcGatewayEndpointOptions {
  /**
   * The VPC network in which the gateway endpoint will be used.
   */
  vpc: IVpcNetwork
}

/**
 * A VPC gateway endpoint.
 */
export class VpcGatewayEndpoint extends cdk.Construct implements IVpcGatewayEndpoint {
  /**
   * Imports an existing gateway endpoint.
   */
  public static import(scope: cdk.Construct, id: string, props: VpcGatewayEndpointImportProps): IVpcGatewayEndpoint {
    return new ImportedVpcGatewayEndpoint(scope, id, props);
  }

  /**
   * The VPC gateway endpoint identifier.
   */
  public readonly vpcEndpointId: string;

  /**
   * The date and time the VPC gateway endpoint was created.
   */
  public readonly creationTimestamp: string;

  private policyDocument?: iam.PolicyDocument;

  constructor(scope: cdk.Construct, id: string, props: VpcGatewayEndpointProps) {
    super(scope, id);

    if (props.service.type !== VpcEndpointType.Gateway) {
      throw new Error('The service endpoint type must be `Gateway`');
    }

    let subnets: IVpcSubnet[] = [];
    if (props.subnets) {
      for (const selection of props.subnets) {
        subnets.push(...props.vpc.subnets(selection));
      }
    } else {
      subnets = props.vpc.subnets();
    }

    const routeTableIds = (subnets as VpcSubnet[]).map(subnet => subnet.routeTableId);

    const endpoint = new CfnVPCEndpoint(this, 'Resource', {
      policyDocument: new cdk.Token(() => this.policyDocument),
      routeTableIds,
      serviceName: props.service.name,
      vpcEndpointType: VpcEndpointType.Gateway,
      vpcId: props.vpc.vpcId
    });

    this.vpcEndpointId = endpoint.vpcEndpointId;
    this.creationTimestamp = endpoint.vpcEndpointCreationTimestamp;
  }

  /**
   * Adds a statement to the policy document.
   *
   * @param statement the statement to add
   */
  public addToPolicy(statement: iam.PolicyStatement) {
    if (!statement.hasPrincipal) {
      throw new Error('Statement must have a `Principal`.');
    }

    if (!this.policyDocument) {
      this.policyDocument = new iam.PolicyDocument();
    }

    this.policyDocument.addStatement(statement);
  }

  /**
   * Exports this VPC gateway endpoint from the stack.
   */
  public export(): VpcGatewayEndpointImportProps {
    return {
      vpcEndpointId: new cdk.CfnOutput(this, 'VpcEndpointId', { value: this.vpcEndpointId }).makeImportValue().toString()
    };
  }
}

/**
 * Construction properties for an ImportedVpcGatewayEndpoint.
 */
export interface VpcGatewayEndpointImportProps {
  /**
   * The VPC gateway endpoint identifier.
   */
  vpcEndpointId: string;
}

/**
 * An imported VPC gateway endpoint.
 */
class ImportedVpcGatewayEndpoint extends cdk.Construct implements IVpcGatewayEndpoint {
  /**
   * The VPC gateway endpoint identifier.
   */
  public readonly vpcEndpointId: string;

  constructor(scope: cdk.Construct, id: string, private readonly props: VpcGatewayEndpointImportProps) {
    super(scope, id);

    this.vpcEndpointId = props.vpcEndpointId;
  }

  /**
   * Exports this VPC gateway endpoint from the stack.
   */
  public export(): VpcGatewayEndpointImportProps {
    return this.props;
  }
}

/**
 * Options to add an interface endpoint to a VPC.
 */
export interface VpcInterfaceEndpointOptions extends VpcEndpointOptions {
  /**
   * Whether to associate a private hosted zone with the specified VPC. This
   * allows you to make requests to the service using its default DNS hostname.
   *
   * @default true
   */
  privateDnsEnabled?: boolean;

  /**
   * The subnets in which to create an endpoint network interface. One per
   * availability zone.
   */
  subnets?: SubnetSelection;
}

/**
 * Construction properties for a VpcInterfaceEndpoint.
 */
export interface VpcInterfaceEndpointProps extends VpcInterfaceEndpointOptions {
  /**
   * The VPC network in which the endpoint will be used.
   */
  vpc: IVpcNetwork
}

/**
 * A VPC interface endpoint.
 */
export interface IVpcInterfaceEndpoint extends IVpcEndpoint, IConnectable {
  /**
   * Exports this VPC interface endpoint from the stack.
   */
  export(): VpcInterfaceEndpointImportProps;
}

/**
 * A VPC interface endpoint.
 */
export class VpcInterfaceEndpoint extends cdk.Construct implements IVpcInterfaceEndpoint {
  /**
   * Imports an existing interface endpoint.
   */
  public static import(scope: cdk.Construct, id: string, props: VpcInterfaceEndpointImportProps): IVpcInterfaceEndpoint {
    return new ImportedVpcInterfaceEndpoint(scope, id, props);
  }

  /**
   * The VPC interface endpoint identifier.
   */
  public readonly vpcEndpointId: string;

  /**
   * The date and time the VPC interface endpoint was created.
   */
  public readonly creationTimestamp: string;

  /**
   * The DNS entries for the VPC interface endpoint.
   */
  public readonly dnsEntries: string[];

  /**
   * One or more network interfaces for the VPC interface endpoint.
   */
  public readonly networkInterfaceIds: string[];

  /**
   * The identifier of the security group associated with this VPC interface
   * endpoint.
   */
  public readonly securityGroupId: string;

  /**
   * Access to network connections.
   */
  public readonly connections: Connections;

  constructor(scope: cdk.Construct, id: string, props: VpcInterfaceEndpointProps) {
    super(scope, id);

    if (props.service.type !== VpcEndpointType.Interface) {
      throw new Error('The service endpoint type must be `Interface`');
    }

    const securityGroup = new SecurityGroup(this, 'SecurityGroup', {
      vpc: props.vpc
    });
    this.securityGroupId = securityGroup.securityGroupId;
    this.connections = new Connections({ securityGroups: [securityGroup] });

    const subnets = props.vpc.subnets(props.subnets);

    const availabilityZones = new Set(subnets.map(s => s.availabilityZone));

    if (availabilityZones.size !== subnets.length) {
      throw new Error('Only one subnet per availability zone is allowed.');
    }

    const endpoint = new CfnVPCEndpoint(this, 'Resource', {
      privateDnsEnabled: props.privateDnsEnabled || true,
      securityGroupIds: [this.securityGroupId],
      serviceName: props.service.name,
      vpcEndpointType: VpcEndpointType.Interface,
      subnetIds: subnets.map(s => s.subnetId),
      vpcId: props.vpc.vpcId
    });

    this.vpcEndpointId = endpoint.vpcEndpointId;
    this.creationTimestamp = endpoint.vpcEndpointCreationTimestamp;
    this.dnsEntries = endpoint.vpcEndpointDnsEntries;
    this.networkInterfaceIds = endpoint.vpcEndpointNetworkInterfaceIds;
  }

  /**
   * Exports this VPC interface endpoint from the stack.
   */
  public export(): VpcInterfaceEndpointImportProps {
    return {
      vpcEndpointId: new cdk.CfnOutput(this, 'VpcEndpointId', { value: this.vpcEndpointId }).makeImportValue().toString(),
      securityGroupId: new cdk.CfnOutput(this, 'SecurityGroupId', { value: this.securityGroupId }).makeImportValue().toString()
    };
  }
}

/**
 * Construction properties for an ImportedVpcInterfaceEndpoint.
 */
export interface VpcInterfaceEndpointImportProps {
  /**
   * The VPC interface endpoint identifier.
   */
  vpcEndpointId: string;

  /**
   * The identifier of the security group associated with the VPC interface endpoint.
   */
  securityGroupId: string;
}

/**
 * An imported VPC interface endpoint.
 */
class ImportedVpcInterfaceEndpoint extends cdk.Construct implements IVpcInterfaceEndpoint {
  /**
   * The VPC interface endpoint identifier.
   */
  public readonly vpcEndpointId: string;

  /**
   * The identifier of the security group associated with the VPC interface endpoint.
   */
  public readonly securityGroupId: string;

  /**
   * Access to network connections.
   */
  public readonly connections: Connections;

  constructor(scope: cdk.Construct, id: string, private readonly props: VpcInterfaceEndpointImportProps) {
    super(scope, id);

    this.vpcEndpointId = props.vpcEndpointId;

    this.securityGroupId = props.securityGroupId;

    this.connections = new Connections({
      securityGroups: [SecurityGroup.import(this, 'SecurityGroup', props)],
    });
  }

  /**
   * Exports this VPC endpoint from the stack.
   */
  public export(): VpcInterfaceEndpointImportProps {
    return this.props;
  }
}
