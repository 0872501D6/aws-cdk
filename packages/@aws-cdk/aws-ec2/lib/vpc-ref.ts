import { Construct, IConstruct, IDependable } from "@aws-cdk/cdk";
import { DEFAULT_SUBNET_NAME, subnetName } from './util';
import { InterfaceVpcEndpoint, InterfaceVpcEndpointOptions } from './vpc-endpoint';
import { VpnConnection, VpnConnectionOptions } from './vpn';

export interface IVpcSubnet extends IConstruct {
  /**
   * The Availability Zone the subnet is located in
   */
  readonly availabilityZone: string;

  /**
   * The subnetId for this particular subnet
   */
  readonly subnetId: string;

  /**
   * Dependable that can be depended upon to force internet connectivity established on the VPC
   */
  readonly internetConnectivityEstablished: IDependable;

  /**
   * Exports this subnet to another stack.
   */
  export(): VpcSubnetImportProps;
}

export interface IVpcNetwork extends IConstruct {
  /**
   * Identifier for this VPC
   */
  readonly vpcId: string;

  /**
   * List of public subnets in this VPC
   */
  readonly publicSubnets: IVpcSubnet[];

  /**
   * List of private subnets in this VPC
   */
  readonly privateSubnets: IVpcSubnet[];

  /**
   * List of isolated subnets in this VPC
   */
  readonly isolatedSubnets: IVpcSubnet[];

  /**
   * AZs for this VPC
   */
  readonly availabilityZones: string[];

  /**
   * Region where this VPC is located
   */
  readonly vpcRegion: string;

  /**
   * Identifier for the VPN gateway
   */
  readonly vpnGatewayId?: string;

  /**
   * Return IDs of the subnets appropriate for the given selection strategy
   *
   * Requires that at least once subnet is matched, throws a descriptive
   * error message otherwise.
   *
   * Prefer to use this method over {@link subnets} if you need to pass subnet
   * IDs to a CloudFormation Resource.
   */
  subnetIds(selection?: SubnetSelection): string[];

  /**
   * Return the subnets appropriate for the placement strategy
   */
  selectSubnets(selection?: SubnetSelection): IVpcSubnet[];

  /**
   * Return a dependable object representing internet connectivity for the given subnets
   */
  subnetInternetDependencies(selection?: SubnetSelection): IDependable;

  /**
   * Return whether all of the given subnets are from the VPC's public subnets.
   */
  isPublicSubnets(subnetIds: string[]): boolean;

  /**
   * Adds a new VPN connection to this VPC
   */
  addVpnConnection(id: string, options: VpnConnectionOptions): VpnConnection;

  /**
   * Adds a new interface endpoint to this VPC
   */
  addInterfaceEndpoint(id: string, options: InterfaceVpcEndpointOptions): InterfaceVpcEndpoint

  /**
   * Exports this VPC so it can be consumed by another stack.
   */
  export(): VpcNetworkImportProps;
}

/**
 * The type of Subnet
 */
export enum SubnetType {
  /**
   * Isolated Subnets do not route Outbound traffic
   *
   * This can be good for subnets with RDS or
   * Elasticache endpoints
   */
  Isolated = 1,

  /**
   * Subnet that routes to the internet, but not vice versa.
   *
   * Instances in a private subnet can connect to the Internet, but will not
   * allow connections to be initiated from the Internet.
   *
   * Outbound traffic will be routed via a NAT Gateway. Preference being in
   * the same AZ, but if not available will use another AZ (control by
   * specifing `maxGateways` on VpcNetwork). This might be used for
   * experimental cost conscious accounts or accounts where HA outbound
   * traffic is not needed.
   */
  Private = 2,

  /**
   * Subnet connected to the Internet
   *
   * Instances in a Public subnet can connect to the Internet and can be
   * connected to from the Internet as long as they are launched with public
   * IPs (controlled on the AutoScalingGroup or other constructs that launch
   * instances).
   *
   * Public subnets route outbound traffic via an Internet Gateway.
   */
  Public = 3
}

/**
 * Customize subnets that are selected for placement of ENIs
 *
 * Constructs that allow customization of VPC placement use parameters of this
 * type to provide placement settings.
 *
 * By default, the instances are placed in the private subnets.
 */
export interface SubnetSelection {
  /**
   * Place the instances in the subnets of the given types
   *
   * At most one of `subnetType` and `subnetName` can be supplied.
   *
   * @default SubnetType.Private
   */
  readonly subnetTypes?: SubnetType[];

  /**
   * Place the instances in the subnets with the given names
   *
   * (This is the name supplied in subnetConfiguration).
   *
   * At most one of `subnetType` and `subnetName` can be supplied.
   *
   * @default name
   */
  readonly subnetNames?: string[];
}

/**
 * A new or imported VPC
 */
export abstract class VpcNetworkBase extends Construct implements IVpcNetwork {

  /**
   * Identifier for this VPC
   */
  public abstract readonly vpcId: string;

  /**
   * List of public subnets in this VPC
   */
  public abstract readonly publicSubnets: IVpcSubnet[];

  /**
   * List of private subnets in this VPC
   */
  public abstract readonly privateSubnets: IVpcSubnet[];

  /**
   * List of isolated subnets in this VPC
   */
  public abstract readonly isolatedSubnets: IVpcSubnet[];

  /**
   * AZs for this VPC
   */
  public abstract readonly availabilityZones: string[];

  /**
   * Identifier for the VPN gateway
   */
  public abstract readonly vpnGatewayId?: string;

  /**
   * Dependencies for internet connectivity
   */
  public readonly internetDependencies = new Array<IConstruct>();

  /**
   * Dependencies for NAT connectivity
   */
  public readonly natDependencies = new Array<IConstruct>();

  /**
   * Returns IDs of selected subnets
   */
  public subnetIds(selection: SubnetSelection = {}): string[] {
    selection = reifySelectionDefaults(selection);

    const nets = this.selectSubnets(selection);
    if (nets.length === 0) {
      throw new Error(`There are no ${describeSelection(selection)} in this VPC. Use a different VPC subnet selection.`);
    }

    return nets.map(n => n.subnetId);
  }

  /**
   * Return a dependable object representing internet connectivity for the given subnets
   */
  public subnetInternetDependencies(selection: SubnetSelection = {}): IDependable {
    selection = reifySelectionDefaults(selection);

    const ret = new CompositeDependable();
    for (const subnet of this.selectSubnets(selection)) {
      ret.add(subnet.internetConnectivityEstablished);
    }
    return ret;
  }

  /**
   * Adds a new VPN connection to this VPC
   */
  public addVpnConnection(id: string, options: VpnConnectionOptions): VpnConnection {
    return new VpnConnection(this, id, {
      vpc: this,
      ...options
    });
  }

  /**
   * Adds a new interface endpoint to this VPC
   */
  public addInterfaceEndpoint(id: string, options: InterfaceVpcEndpointOptions): InterfaceVpcEndpoint {
    return new InterfaceVpcEndpoint(this, id, {
      vpc: this,
      ...options
    });
  }

  /**
   * Export this VPC from the stack
   */
  public abstract export(): VpcNetworkImportProps;

  /**
   * Return whether all of the given subnets are from the VPC's public subnets.
   */
  public isPublicSubnets(subnetIds: string[]): boolean {
    const pubIds = new Set(this.publicSubnets.map(n => n.subnetId));
    return subnetIds.every(pubIds.has.bind(pubIds));
  }

  /**
   * The region where this VPC is defined
   */
  public get vpcRegion(): string {
    return this.node.stack.region;
  }

  /**
   * Return the subnets appropriate for the placement strategy
   */
  public selectSubnets(selection: SubnetSelection = {}): IVpcSubnet[] {
    selection = reifySelectionDefaults(selection);

    // Select by name
    if (selection.subnetNames !== undefined) {
      const allSubnets = [...this.publicSubnets, ...this.privateSubnets, ...this.isolatedSubnets];
      const names = selection.subnetNames;
      const nameSubnets = allSubnets.filter(s => names.includes(subnetName(s)));
      if (nameSubnets.length === 0) {
        throw new Error(`No subnets with names in: ${selection.subnetNames}`);
      }
      return nameSubnets;
    }

    // Select by type
    if (selection.subnetTypes === undefined) { return this.privateSubnets; }

    let typeSubnets: IVpcSubnet[] = [];
    if (selection.subnetTypes.includes(SubnetType.Public)) {
      typeSubnets = [...typeSubnets, ...this.publicSubnets];
    }
    if (selection.subnetTypes.includes(SubnetType.Private)) {
      typeSubnets = [...typeSubnets, ...this.privateSubnets];
    }
    if (selection.subnetTypes.includes(SubnetType.Isolated)) {
      typeSubnets = [...typeSubnets, ...this.isolatedSubnets];
    }
    return typeSubnets;
  }
}

/**
 * Properties that reference an external VpcNetwork
 */
export interface VpcNetworkImportProps {
  /**
   * VPC's identifier
   */
  readonly vpcId: string;

  /**
   * List of availability zones for the subnets in this VPC.
   */
  readonly availabilityZones: string[];

  /**
   * List of public subnet IDs
   *
   * Must be undefined or match the availability zones in length and order.
   */
  readonly publicSubnetIds?: string[];

  /**
   * List of names for the public subnets
   *
   * Must be undefined or have a name for every public subnet group.
   */
  readonly publicSubnetNames?: string[];

  /**
   * List of private subnet IDs
   *
   * Must be undefined or match the availability zones in length and order.
   */
  readonly privateSubnetIds?: string[];

  /**
   * List of names for the private subnets
   *
   * Must be undefined or have a name for every private subnet group.
   */
  readonly privateSubnetNames?: string[];

  /**
   * List of isolated subnet IDs
   *
   * Must be undefined or match the availability zones in length and order.
   */
  readonly isolatedSubnetIds?: string[];

  /**
   * List of names for the isolated subnets
   *
   * Must be undefined or have a name for every isolated subnet group.
   */
  readonly isolatedSubnetNames?: string[];

  /**
   * VPN gateway's identifier
   */
  readonly vpnGatewayId?: string;
}

export interface VpcSubnetImportProps {
  /**
   * The Availability Zone the subnet is located in
   */
  readonly availabilityZone: string;

  /**
   * The subnetId for this particular subnet
   */
  readonly subnetId: string;
}

/**
 * If the placement strategy is completely "default", reify the defaults so
 * consuming code doesn't have to reimplement the same analysis every time.
 *
 * Returns "private subnets" by default.
 */
function reifySelectionDefaults(placement: SubnetSelection): SubnetSelection {
    if (placement.subnetTypes !== undefined && placement.subnetNames !== undefined) {
      throw new Error('Only one of subnetType and subnetName can be supplied');
    }

    if (placement.subnetTypes === undefined && placement.subnetNames === undefined) {
      return { subnetTypes: [SubnetType.Private] };
    }

    return placement;
}

/**
 * Describe the given placement strategy
 */
function describeSelection(placement: SubnetSelection): string {
  if (placement.subnetTypes !== undefined) {
    return `'${joinCommaOr(placement.subnetTypes.map(type => DEFAULT_SUBNET_NAME[type]))}' subnets`;
  }
  if (placement.subnetNames !== undefined) {
    return `subnets named '${joinCommaOr(placement.subnetNames)}'`;
  }
  return JSON.stringify(placement);
}

function joinCommaOr(array: string[]) {
  return [array.slice(0, -1).join(', '), array.slice(-1)[0]].join(array.length < 2 ? '' : ' or ');
}

class CompositeDependable implements IDependable {
  private readonly dependables = new Array<IDependable>();

  /**
   * Add a construct to the dependency roots
   */
  public add(dep: IDependable) {
    this.dependables.push(dep);
  }

  /**
   * Retrieve the current set of dependency roots
   */
  public get dependencyRoots(): IConstruct[] {
    const ret = [];
    for (const dep of this.dependables) {
      ret.push(...dep.dependencyRoots);
    }
    return ret;
  }
}
