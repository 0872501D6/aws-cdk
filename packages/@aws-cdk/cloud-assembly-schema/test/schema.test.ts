import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { AssemblyManifest, Manifest, StackTagsMetadataEntry } from '../lib';
import { hashObject } from './fingerprint';

const FIXTURES = path.join(__dirname, 'fixtures');

test('manifest save', () => {

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'protocol-tests'));
  const manifestFile = path.join(outdir, 'manifest.json');

  const assemblyManifest: AssemblyManifest = {
    version: "version"
  };

  Manifest.save(assemblyManifest, manifestFile);

  const saved = JSON.parse(fs.readFileSync(manifestFile, 'UTF-8'));

  expect(saved).toEqual(assemblyManifest);

});

test('manifest load', () => {

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'protocol-tests'));
  const manifestFile = path.join(outdir, 'manifest.json');

  const assemblyManifest: AssemblyManifest = {
    version: "version"
  };

  Manifest.save(assemblyManifest, manifestFile);

  const loaded = Manifest.load(manifestFile);

  expect(loaded).toEqual(assemblyManifest);

});

test('manifest load fail on invalid file', () => {

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'protocol-tests'));
  const manifestFile = path.join(outdir, 'manifest.json');

  // this is invalid because 'version' is required
  const assemblyManifest = {

  };

  fs.writeFileSync(manifestFile, JSON.stringify(assemblyManifest));

  expect(() => Manifest.load(manifestFile)).toThrow(/Invalid assembly manifest/);

});

test('schema has the correct hash', () => {

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const expectedHash = require('./schema.expected.json').hash;

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const schema = require('../schema/cloud-assembly.schema.json');

  const schemaHash = hashObject(schema);

  expect(schemaHash).toEqual(expectedHash);

});

test('manifest load fail on invalid file', () => {

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'protocol-tests'));
  const manifestFile = path.join(outdir, 'manifest.json');

  // this is invalid because 'version' is required
  const assemblyManifest = {

  };

  fs.writeFileSync(manifestFile, JSON.stringify(assemblyManifest));

  expect(() => Manifest.load(manifestFile)).toThrow(/Invalid assembly manifest/);

});

test('manifest load fail on complex invalid file', () => {

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'protocol-tests'));
  const manifestFile = path.join(outdir, 'manifest.json');

  // this is invalid because 'version' is required
  const assemblyManifest = {
    version: "0.0.5",
    runtime: {
      libraries: ["should", "be", "a", "map"]
    }
  };

  fs.writeFileSync(manifestFile, JSON.stringify(assemblyManifest));

  expect(() => Manifest.load(manifestFile)).toThrow(/Invalid assembly manifest/);

});

test('load fails for invalid artifact type', () => {
  expect(() => Manifest.load(path.join(FIXTURES, 'invalid-artifact-type', 'manifest.json')))
  .toThrow(/Invalid assembly manifest/);
});

test('stack-tags are deserialized properly', () => {

  const outdir = fs.mkdtempSync(path.join(os.tmpdir(), 'protocol-tests'));
  const manifestFile = path.join(outdir, 'manifest.json');

  fs.writeFileSync(manifestFile, JSON.stringify({
    version: "version",
    artifacts: {
      Tree: {
        type: "cdk:tree",
        properties: {
          file: "tree.json"
        }
      },
      stack: {
        type: "aws:cloudformation:stack",
        metadata: {
          AwsCdkPlaygroundBatch: [
            {
              type: "aws:cdk:stack-tags",
              data: [{
                Key: "hello",
                Value: "world"
              }],
              trace: ["trace"]
            },
            {
              type: "aws:cdk:asset",
              data: {
                repositoryName: "repo",
                imageTag: "tag",
                id: "id",
                packaging: "container-image",
                path: "path",
                sourceHash: "hash"
              },
              trace: ["trace"]
            },
          ]
        }
      }
    },
  }));

  const m: AssemblyManifest = Manifest.load(manifestFile);

  if (m.artifacts?.stack?.metadata?.AwsCdkPlaygroundBatch[0].data) {
    const entry = m.artifacts.stack.metadata.AwsCdkPlaygroundBatch[0].data as StackTagsMetadataEntry;
    expect(entry[0].key).toEqual("hello");
    expect(entry[0].value).toEqual("world");
  }
  expect(m.version).toEqual("version");

});
