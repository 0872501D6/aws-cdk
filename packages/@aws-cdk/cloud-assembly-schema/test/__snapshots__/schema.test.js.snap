// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manifest load 1`] = `
Object {
  "version": "version",
}
`;

exports[`metadata is not validated 1`] = `
Object {
  "artifacts": Object {
    "Tree": Object {
      "properties": Object {
        "file": "tree.json",
      },
      "type": "cdk:tree",
    },
    "stack": Object {
      "metadata": Object {
        "AwsCdkPlaygroundBatch": Array [
          Object {
            "data": Array [
              "42",
            ],
            "trace": Array [
              "trace",
            ],
            "type": "something",
          },
          Object {
            "data": 42,
            "trace": Array [
              "trace",
            ],
            "type": "something-else",
          },
        ],
      },
      "type": "aws:cloudformation:stack",
    },
  },
  "version": "version",
}
`;
